### PR TITLE
[REVIEW] Clean up of UCX progress

### PR DESCRIPTION
**Pull request by madsbk**
This PR moves ucx progress in its own file and use classes to simplify the clean up process.

### CODE REVIEW

**Comment by quasiben** [comment]
rerun tests

**Comment by madsbk** [comment]
rerun tests

File: ucp/continuous_ucx_progress.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        event_loop: asyncio.EventLoop
+            The event loop to do progress in.
+        """
+        self.weakref_ctx = weakref.ref(ctx)
+        self.event_loop = event_loop
+        self.asyncio_task = None
+
+    def __del__(self):
+        if self.asyncio_task is not None:
+            self.asyncio_task.cancel()
```
Comment (jakirkham) [inline ucp/continuous_ucx_progress.py]: What happens if the task scopes out normally?

File: ucp/continuous_ucx_progress.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        return hash(self) == hash(other)
+
+
+class NonBlockingMode(ProgressTask):
+    def __init__(self, ctx, event_loop):
+        super().__init__(ctx, event_loop)
+        self.asyncio_task = event_loop.create_task(self._progress_task())
+
+    async def _progress_task(self):
+        """This help function maintains a UCX progress loop."""
```
Comment (jakirkham) [inline ucp/continuous_ucx_progress.py]:
```suggestion
        """This helper function maintains a UCX progress loop."""
```

File: ucp/_libs/core.pyx [hunk 3]
```diff
@@ -533,75 +513,32 @@ cdef class ApplicationContext:
...
         else:
-            self._non_blocking_progress_mode(loop)
+            task = continuous_ucx_progress.NonBlockingMode(self, loop)
+        self.progress_tasks.append(task)
 
     def get_ucp_worker(self):
         return int(<uintptr_t><void*>self.worker)
 
     def get_config(self):
         return self.config
```
Comment (jakirkham) [inline ucp/_libs/core.pyx]: Trimming some whitespace 🙂

```suggestion
```

File: ucp/_libs/core.pyx [hunk 4]
```diff
@@ -411,10 +394,7 @@ cdef class ApplicationContext:
                 child = child()
                 if child is not None:
                     child.abort()
-            for task in (self.progress_tasks if self.progress_tasks else []):
-                task.cancel()
-            if self.dangling_arm_task is not None:
-                self.dangling_arm_task.cancel()
+            self.progress_tasks = None
```
Comment (jakirkham) [inline ucp/_libs/core.pyx]: Should this be assigned an empty `list`?

```suggestion
            self.progress_tasks = []
```

File: ucp/continuous_ucx_progress.py (hunk 1 above)
Comment (madsbk) [reply ucp/continuous_ucx_progress.py]: That is fine, when this goes out of scope the continuous progress will stop

Comment (jakirkham) [reply ucp/continuous_ucx_progress.py]: Ok, am just wondering do we need to call `.cancel()` explicitly or does this already happens for us?

Comment (madsbk) [reply ucp/continuous_ucx_progress.py]: In some cases we have to call it explicitly.
PS: note that it is fine to call `.cancel()` even if the future is already done.

**Comment by quasiben** [comment]
Thanks @madsbk and thank you for the review @jakirkham

**Pull request merged**